docs/design/specs/Memory-model.md

```diff
@@ -0,0 +1,276 @@
+     - `System.Threading.Volatile.Write`
+     - `System.Thread.VolatileWrite`
+     - Releasing a lock (`System.Threading.Monitor.Exit` or leaving a synchronized method)
+
+* **volatile. initblk** has "release semantics" - the effects of `.volatile initblk` will not be observable earlier than the effects of preceeding reads and writes.
+
+* **volatile. cpblk** combines ordering semantics of a volatile read and write with respect to the read and written memory locations.
+     - The writes performed by `volatile. cpblk` will not be observable earlier than the effects of preceeding reads and writes.
+     - No read or write that is later in the program order may be speculatively executed before the reads performed by `volatile. cpblk`
+     - `cpblk` may be implemented as a sequence of reads and writes. The granularity and mutual order of such reads and writes is unspecified.
+
+Note that volatile semantics does not by itself imply that operation is atomic or has any effect on how soon the operation is committed to the coherent memory. It only specifies the order of effects when they eventually become observable.
+
+`volatile.` and `unaligned.` IL prefixes can be combined where both are permitted.
+
+It may be possible for an optimizing compiler to prove that some data is accessible only by a single thread. In such case it is permitted to omit volatile semantics when accessing such data.
+
+* **Full-fence operations**
+  Full-fence operations have "full-fence semantics" - effects of reads and writes must be observable no later or no earlier than a full-fence operation according to their relative program order.
+  Operations with full-fence semantics:
+     - `System.Thread.MemoryBarrier`
+     - `System.Threading.Interlocked` methods
+
+## C# `volatile` feature
+One common way to introduce volatile memory accesses is by using C# `volatile` language feature. Declaring a field as `volatile` does not have any effect on how .NET runtime treats the field. The decoration works as a hint to the C# compiler itself (and compilers for other .Net languages) to emit reads and writes of such field as  reads and writes with `volatile.` prefix.
+
+## Process-wide barrier
+Process-wide barrier has full-fence semantics with an additional guarantee that each thread in the program effectively performs a full fence at arbitrary point synchronized with the process-wide barrier in such a way that effects of writes that precede both barriers are observable by memory operations that follow the barriers.
+
+The actual implementation may vary depending on the platform. For example interrupting the execution of every core in the current process' affinity mask could be a suitable implementation.
+
+## Synchronized methods
+Methods decorated with ```MethodImpl(MethodImplOptions.Synchronized)``` attribute have the same memory access semantics as if a lock is acquired at an entrance to the method and released upon leaving the method.
+
+## Data-dependent reads are ordered
```

Comment:
The way specifying the data dependency in th ehardware:

- for reads we do not need to do anything. on all supported platforms dependent reads are ordered. (val = X.Y)
if a value from location X was used to compute address Y, reading from Y cannot happen earlier than reading from X; i.e. we cannot get a cached value of a field that is older than the value of the instance. Note - it is ok if both were cached long time ago, as long as value of Y is not older than X.

Compiler will typically not violate this, since it would not know how to read Y without reading X. However with some prefetching, or profile guided optimization, it might in theory be possible to "know" the location of Y and read it directly (not via X), so we forbid such optimizations in this doc.

- for writes we need to use a release store. On x64 all stores are releases, but on ARM64 we need to use STLR, which ensures that results of the store will not be observable earlier than any memory operation that precedes it (in program order).

```
x.a = val1;
x.b = a;
// we emit STLR here, effects of all memory operations above will become observable no later than the write below.
nonlocal = x; 
```

I think we need to forbid compiler optimizations that violate such guarantee. Otherwise there is no point to instruct the hardware to do the ordering.

```
x.y = val;
nonlocal = x;
```

the following emit should be invalid
```
nonlocal = x;
x.y = val;
```

Just saying "nonlocal writes cannot reorder" would prevent reordering of unrelated writes, which seems a bit too strict.

```
nonlocal1 = val1;
nonlocal2 = val2;
```
should be, in theory, ok to emit as: 
```
nonlocal2 = val2;
nonlocal1 = val1;
```

In a case of complex aliasing, compiler could just behave conservatively and keep the program order of writes, if the writes are not local and not known to be "unrelated".

I am just not sure how to nicely define "related"  -  reachable through an indirection, through a dependent read?

Reply:
Not sure about all this. I suspect the jit will not do store reordering today, but I can't say for sure.

Dead stores may also be a problem, say we have
```
x.y = val
nonlocal = x
x.y = val1
```
and the jit can prove `nonlocal` and `x` don't alias, I think it may remove the first store as dead. But maybe this only happens when `x` is known to be a stack location.

In the "transitive case"
```
x.y = val
nonlocal = z  // root of some object graph that reaches x
```
presumably `x` has already been "exposed" in some way?

Reply:
```
x.y = val
nonlocal = x
x.y = val1
```

Removing the first write should be illegal in this case. If `x.y` starts as `null` and `nonlocal = x` is the first publish, the reader might expect that `nonlocal?.y.ToString()` will not throw NRE.

The following seems ok:

```
x.y = val
x.y = val1
nonlocal = x
```

```
x.y = val1
nonlocal = x
```

If publishing is a release, it cannot allow preceding memory operations to be delayed after it, but it would be ok for the operations after it to move earlier.

If we instruct the hardware do a release, I think the compiler should follow the same rules.

Perhaps we can just document it as something like - "for the purpose of compiler optimizations nonlocal object writes should be treated as releases".  

We may not always emit the assignment as an actual release. For example when assigning a frozen/preallocated string literal, we do not need a release in hardware, since the string is immutable and is guaranteed to be completely formed, but we still must preserve the order of writes in compiler optimizations.